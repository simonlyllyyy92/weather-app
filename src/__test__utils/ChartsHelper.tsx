import * as d3 from "d3";

//Type from context
import { DataType } from "../context/set_context";

/**
 * @Function This is a similar charts generate function But for testing purposes
 * @param mappedData  Data from Mock Service work will be passed in
 */

export const chartsGenerate = (mappedData: DataType[]) => {
  /**
   * For testing purposes we set up fix width and height here
   */
  const currentWidth = 600;
  const currentHeight = 400;

  //Setting default svg height and width
  const margin = { top: 28, right: 20, bottom: 30, left: 20 };
  const width = currentWidth - margin.left - margin.right;
  const height = currentHeight - margin.top - margin.bottom;
  /**
   * @return Maximum temperature value and Minimum temperature value
   */
  const getDomain = (rows: any) => {
    const temp_val = rows.map(({ value }: { value: number }) => [value]).flat();
    return [Math.min(...temp_val), Math.max(...temp_val)];
  };
  /**
   * @return  Maximum date value and Minimum Date Value
   */

  const getTimeStampDomain = (rows: any) => {
    const value = rows.map(({ date }: { date: number }) => [date]).flat();
    return [Math.min(...value), Math.max(...value)];
  };

  //Get the domain value for date and temperature
  const domain = getDomain(mappedData);
  const time_domain = getTimeStampDomain(mappedData);

  //Set up svg image height and width
  const svg = d3
    .select("#d3_container")
    .append("svg")
    .attr("width", currentWidth)
    .attr("height", currentHeight);

  // Make sure each time the page reload, the svg will be cleared !important
  svg.selectAll("*").remove();

  /**
   * Adjust the axis position
   * As some of the label on the axies might not be displayed properly
   */
  const axisWidth = width - 20;
  const axisHeight = height + 20;

  //Set up d3 scale function
  const xScale = d3.scaleTime().domain(time_domain).range([0, axisWidth]);
  const yScale = d3.scaleLinear().domain(domain).range([height, 0]);

  //Generate Axis X
  svg
    .append("g")
    .call(d3.axisBottom(xScale).tickSizeOuter(0).ticks(6))
    .attr("color", "#000")
    .attr("transform", `translate(20,${axisHeight})`);
  //Generate Axis Y
  svg
    .append("g")
    .call(d3.axisRight(yScale).tickSizeInner(2).tickSizeOuter(0).ticks(8))
    .attr("color", "#000")
    .attr("transform", `translate(${axisWidth + 20}, 20)`);
  //Generate Text hint
  svg
    .append("text")
    .attr("fill", "#3C4E76")
    .attr("transform", `translate(${width - 12}, 10)`)
    .attr("y", 6)
    .attr("dy", "0.71em")
    .attr("text-anchor", "end")
    .text("Temperature (Â°C)");

  // Draw area
  const areaGenerator = d3.area<DataType>().y0(height);

  areaGenerator
    .x(function (d, i) {
      return xScale(d.date);
    })
    .y1(function (d) {
      return yScale(d.value);
    });
  const area = areaGenerator(mappedData);

  svg
    .append("path")
    .attr("fill", "#3C4E76")
    .attr("stroke", "#999")
    .attr("transform", `translate(20,20)`)
    .attr("d", area);

  return svg;
};
