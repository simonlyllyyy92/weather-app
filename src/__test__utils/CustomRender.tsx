import { render } from "@testing-library/react";
import { WeatherContextProvider } from "../context/set_context";

export const customRender = (ui: JSX.Element, options?: any) => {
  return render(ui, { wrapper: WeatherContextProvider, ...options });
};
