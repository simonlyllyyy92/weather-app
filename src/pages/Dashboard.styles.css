.card_container {
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column;
  height: 100vh;
  width: 100vw;
}

.app_header_container {
  align-self: flex-start;
  margin-left: calc(100vw - 95%);
  margin-bottom: 24px;
}
.app_header_container .web_header {
  color: white;
  font-size: 32px;
}
.app_header_container .web_sub_header {
  color: white;
  font-size: 16px;
}

.card_box {
  display: flex;
  width: 90%;
  height: 70%;
  flex-direction: row;
  box-shadow: rgba(0, 0, 0, 0.25) 0px 14px 28px,
    rgba(0, 0, 0, 0.22) 0px 10px 10px;
}

/*
*@Left card style
*/
.card_left_welcome {
  width: 70%;
  height: 100%;
  background: url("../assets/weathorIcon/images/ent_background.png") center
    center;
  background-repeat: no-repeat;
  background-size: cover;
}
.card_left {
  width: 70%;
  height: 100%;
  background: url("../assets/weathorIcon/images/cloudy_background.gif") top
    center;
  background-repeat: no-repeat;
  background-size: cover;
  display: flex;
  align-items: center;
  justify-content: center;
  object-fit: cover;
}

.card_left .card_left_content_box {
  border: 1px solid red;
  height: 80%;
  width: 90%;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
}
.card_left .card_left_charts_box {
  width: 100%;
  height: 100%;
  display: flex;
  justify-content: center;
  border: 1px solid red;
  align-items: center;
}

.card_left_charts_box .card_left_charts {
  width: 80%;
  height: 80%;
  display: flex;
  justify-content: center;
  align-items: center;
  border: 1px solid blue;
  background-color: white;
  opacity: 0.6;
  box-shadow: 4px 8px 8px hsl(0deg 0% 0% / 0.38);
  border-radius: 1em;
}

.card_title {
  display: flex;
  border: 1px solid black;
  align-items: center;
  height: 140px;
}

.card_title .temp {
  font-size: 86px;
  color: white;
}
.card_header .city {
  display: flex;
  align-items: center;
  justify-content: space-between;
}

.city .city_name {
  font-size: 28px;
  color: white;
  margin-right: 12px;
}
.city .city_time {
  font-size: 18px;
  color: white;
  align-self: bottom;
}

.card_title .weather {
  margin-left: 36px;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
}
.iconlabel {
  color: white;
  font-size: 18px;
  text-align: center;
}

/*
*   Right Card style
*/
.card_right {
  width: 30%;
  height: 100%;
  background: url("../assets/weathorIcon/images/sea-blue.png") no-repeat center
    center;
  background-size: cover;
  z-index: 1;
  display: flex;
  justify-content: center;
  align-items: center;
}

.card_right .card_right_content_box {
  border: 1px solid red;
  height: 80%;
  width: 90%;
  display: flex;
  flex-direction: column;
}

.card_right .input_container {
  width: 100%;
  display: flex;
  justify-content: center;
  flex-direction: column;
  align-items: center;
}
