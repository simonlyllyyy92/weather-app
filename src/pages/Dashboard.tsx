import React from "react";
//styles
import "./Dashboard.styles.css";
//Components
import { CardLeft } from "./component/CardLeft";
import { CardRight } from "./component/CardRight";

/**
 * @Dashboard Pages
 * @returns JSX Element
 */

export const Dashboard = (): JSX.Element => {
  return (
    <div className="card_container">
      <div className="app_header_container">
        <div className="web_header" data-testid="app_header">
          Welcome to the Weather Forecast Application
        </div>
        <div className="web_sub_header" data-testid="app_sub_header">
          Please input a city from right to start using this app!
        </div>
      </div>
      <div className="card_box">
        <CardLeft />
        <CardRight />
      </div>
    </div>
  );
};
