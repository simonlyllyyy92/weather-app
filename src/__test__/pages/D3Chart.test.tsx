import { Dashboard } from "../../pages/Dashboard";
import {
  screen,
  waitFor,
  waitForElementToBeRemoved,
} from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { chartsGenerate } from "../../__test__utils/ChartsHelper";
import { customRender } from "../../__test__utils/CustomRender";

test("Charts should be same with the same input", async () => {
  customRender(<Dashboard />);

  // input Sydney
  const city_input = screen.getByTestId("city_input");
  await waitFor(async () => await userEvent.type(city_input, "Sydney"));
  // click submit
  const search_submit = await screen.findByTestId("search_submit");
  userEvent.click(search_submit);

  //welcome page will disappear
  await waitForElementToBeRemoved(() => screen.getByTestId("welcome-card"));

  //this is the demo data created based on MSW handler response
  const mapped_data_from_msw = [
    {
      date: new Date("2021-11-10 00:00"),
      value: 18.3,
    },
    {
      date: new Date("2021-11-10 01:00"),
      value: 19.3,
    },
    {
      date: new Date("2021-11-10 02:00"),
      value: 20.3,
    },
  ];
  //Match snap shot of the generated charts
  // and test based on the same input from MSW, will the same svg be genereated
  const charts = chartsGenerate(mapped_data_from_msw);
  await waitFor(() => expect(charts.node()).toMatchSnapshot());
});
