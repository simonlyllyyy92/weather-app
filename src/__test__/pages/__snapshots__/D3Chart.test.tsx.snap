// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Charts should be same with the same input 1`] = `
<svg
  height="400"
  width="600"
>
  <g
    color="#000"
    fill="none"
    font-family="sans-serif"
    font-size="10"
    text-anchor="middle"
    transform="translate(20,362)"
  >
    <path
      class="domain"
      d="M0.5,0.5H540.5"
      stroke="currentColor"
    />
    <g
      class="tick"
      opacity="1"
      transform="translate(0.5,0)"
    >
      <line
        stroke="currentColor"
        y2="6"
      />
      <text
        dy="0.71em"
        fill="currentColor"
        y="9"
      >
        Wed 10
      </text>
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(68,0)"
    >
      <line
        stroke="currentColor"
        y2="6"
      />
      <text
        dy="0.71em"
        fill="currentColor"
        y="9"
      >
        12:15
      </text>
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(135.5,0)"
    >
      <line
        stroke="currentColor"
        y2="6"
      />
      <text
        dy="0.71em"
        fill="currentColor"
        y="9"
      >
        12:30
      </text>
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(203,0)"
    >
      <line
        stroke="currentColor"
        y2="6"
      />
      <text
        dy="0.71em"
        fill="currentColor"
        y="9"
      >
        12:45
      </text>
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(270.5,0)"
    >
      <line
        stroke="currentColor"
        y2="6"
      />
      <text
        dy="0.71em"
        fill="currentColor"
        y="9"
      >
        01 AM
      </text>
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(338,0)"
    >
      <line
        stroke="currentColor"
        y2="6"
      />
      <text
        dy="0.71em"
        fill="currentColor"
        y="9"
      >
        01:15
      </text>
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(405.5,0)"
    >
      <line
        stroke="currentColor"
        y2="6"
      />
      <text
        dy="0.71em"
        fill="currentColor"
        y="9"
      >
        01:30
      </text>
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(473,0)"
    >
      <line
        stroke="currentColor"
        y2="6"
      />
      <text
        dy="0.71em"
        fill="currentColor"
        y="9"
      >
        01:45
      </text>
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(540.5,0)"
    >
      <line
        stroke="currentColor"
        y2="6"
      />
      <text
        dy="0.71em"
        fill="currentColor"
        y="9"
      >
        02 AM
      </text>
    </g>
  </g>
  <g
    color="#000"
    fill="none"
    font-family="sans-serif"
    font-size="10"
    text-anchor="start"
    transform="translate(560, 20)"
  >
    <path
      class="domain"
      d="M0.5,342.5V0.5"
      stroke="currentColor"
    />
    <g
      class="tick"
      opacity="1"
      transform="translate(0,325.4000000000004)"
    >
      <line
        stroke="currentColor"
        x2="2"
      />
      <text
        dy="0.32em"
        fill="currentColor"
        x="5"
      >
        18.4
      </text>
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(0,291.1999999999999)"
    >
      <line
        stroke="currentColor"
        x2="2"
      />
      <text
        dy="0.32em"
        fill="currentColor"
        x="5"
      >
        18.6
      </text>
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(0,257)"
    >
      <line
        stroke="currentColor"
        x2="2"
      />
      <text
        dy="0.32em"
        fill="currentColor"
        x="5"
      >
        18.8
      </text>
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(0,222.80000000000013)"
    >
      <line
        stroke="currentColor"
        x2="2"
      />
      <text
        dy="0.32em"
        fill="currentColor"
        x="5"
      >
        19.0
      </text>
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(0,188.60000000000025)"
    >
      <line
        stroke="currentColor"
        x2="2"
      />
      <text
        dy="0.32em"
        fill="currentColor"
        x="5"
      >
        19.2
      </text>
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(0,154.40000000000038)"
    >
      <line
        stroke="currentColor"
        x2="2"
      />
      <text
        dy="0.32em"
        fill="currentColor"
        x="5"
      >
        19.4
      </text>
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(0,120.19999999999987)"
    >
      <line
        stroke="currentColor"
        x2="2"
      />
      <text
        dy="0.32em"
        fill="currentColor"
        x="5"
      >
        19.6
      </text>
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(0,86)"
    >
      <line
        stroke="currentColor"
        x2="2"
      />
      <text
        dy="0.32em"
        fill="currentColor"
        x="5"
      >
        19.8
      </text>
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(0,51.800000000000125)"
    >
      <line
        stroke="currentColor"
        x2="2"
      />
      <text
        dy="0.32em"
        fill="currentColor"
        x="5"
      >
        20.0
      </text>
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(0,17.600000000000243)"
    >
      <line
        stroke="currentColor"
        x2="2"
      />
      <text
        dy="0.32em"
        fill="currentColor"
        x="5"
      >
        20.2
      </text>
    </g>
  </g>
  <text
    dy="0.71em"
    fill="#3C4E76"
    text-anchor="end"
    transform="translate(548, 10)"
    y="6"
  >
    Temperature (Â°C)
  </text>
  <path
    d="M0,342L270,171L540,0L540,342L270,342L0,342Z"
    fill="#3C4E76"
    stroke="#999"
    transform="translate(20,20)"
  />
</svg>
`;
